<a href="https://colab.research.google.com/github/MOOwuttichai/BSC_DPDM2023/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC-dpdm23_flie'

Mounted at /content/drive


In [3]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [4]:
data
data_test=data

In [5]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

#Quit1 มีกี่ประเทศ

In [6]:
len(set(data['Country']))

38

#เเปลง data ใก้อยู่ในรูปเเบบ Transaction data ด้วย ```groupby.apply(list) ```

In [7]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [8]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [9]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=621ee7c3617c14ae07b15502c2a7d577e023a1bea6980db6667c7465ba42de5b
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [10]:
from apyori import apriori

In [11]:
!pip install apyori
from apyori import apriori
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [12]:
len(frequentItemsets)

4

In [13]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [14]:
frequentItemsets[0][0]

frozenset({'22423'})

In [15]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [16]:
frequentItemsets[0][0].__len__()#หาขนาดของ itemset

1

In [17]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),
support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [18]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

#HW 1 หา Frequent Item Set เเละ Association Rules
* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)
* ของช่วงเวลา Christmas & New Year ในทวีป Europe(1 Dec - 31 Dec)
* ของวันวาเลนไทน์ (14 Feb)
* จัดกลุ่มตามประเทศ+เดิอน

##1.ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)

###เเยกวันเดือนปีออกจากเวลา

In [19]:
data['InvoiceDate'][0].split()[0]

'12/01/2010'

In [20]:
date =[]
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)
#set(L)

In [21]:
date[0].split('/')

['12', '01', '2010']

###เเยกวัน เดือน ปีออกจากกกัน

In [22]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  day.append(x)
  month.append(y)
  year.append(z)

###ใส่ดาต้าใหม่

In [23]:
data_test['date']=date
data_test['day']=day
data_test['month']=month
data_test['year']=year

In [24]:
data_test

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

       day month  year  
0       12    01  2010  
1       12    01  2010  
2       12    01  2010  
3       12    01  2010  
4       12    01  2010  
...     ..   ...   ...  
541904  12    09  2011  
541905  12    09  2011  
541906  12    09  2011  
541907  12    09  2011  
541908  12    09  2011  

[541909 rows x 12 columns]

###เริ่มทำข้อ 1


In [25]:
data_C_and_N = data_test[data_test['month']=='12']

In [26]:
!pip install apyori
from apyori import apriori



In [27]:
transactionsC_and_N = data_C_and_N.groupby('InvoiceNo')['StockCode'].apply(list)

In [28]:
frequentItemsets = list(apriori(transactionsC_and_N,min_support = 0.05))
#frequentItemsets

In [29]:
min_sup = 0.03

frequentItemsetsC_and_N = list(apriori(transactionsC_and_N,min_support = min_sup))
for i in frequentItemsetsC_and_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22383', '20725'}), support=0.031203566121842496, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22383', '20725'}), confidence=0.031203566121842496, lift=1.0), OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'22383'}), confidence=0.4666666666666667, lift=9.237254901960783), OrderedStatistic(items_base=frozenset({'22383'}), items_add=frozenset({'20725'}), confidence=0.6176470588235294, lift=9.237254901960785)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03417533432392273, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03417533432392273, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7666666666666666, lift=18.42738095238095), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.8214285714285714, lift=18.42

สรุปได้ดังนั้น
* RelationRecord(items=frozenset({'20725', '22383'}), support=0.031203566121842496)
  * ordered_statistics=OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725', '22383'}), confidence=0.031203566121842496, lift=1.0)
  * OrderedStatistic(items_base=frozenset({'20725'}), items_add=frozenset({'22383'}), confidence=0.4666666666666667, lift=9.237254901960783)
  * OrderedStatistic(items_base=frozenset({'22383'}), items_add=frozenset({'20725'}), confidence=0.6176470588235294, lift=9.237254901960785)
* RelationRecord(items=frozenset({'22697', '22699'}), support=0.03417533432392273)
  * ordered_statistics=OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03417533432392273, lift=1.0)
  * OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7666666666666666, lift=18.42738095238095)
  * OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.8214285714285714, lift=18.42738095238095)

##2.ของประเทศในทวีป Europe

##เเบ่งทวีป

In [30]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

ทวีปยุโรป

In [31]:
UP_country = ['Austria','Belgium','Cyprus','Czech Republic','Denmark','EIRE','European Community','Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands','Norway','Poland','Portugal','Spain','Sweden','United Kingdom']

ทวีปเอเชีย

In [32]:
Asia_country = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [42]:
a0=data_test[data_test['Country']==UP_country[0]]

InvoiceNo StockCode                          Description  Quantity  \
7279      537026     84375        SET OF 20 KIDS COOKIE CUTTERS        12   
7280      537026     21217        RED RETROSPOT ROUND CAKE TINS         2   
7281      537026     21212      PACK OF 72 RETROSPOT CAKE CASES       120   
7282      537026     21977   PACK OF 60 PINK PAISLEY CAKE CASES       120   
7283      537026     22417       PACK OF 60 SPACEBOY CAKE CASES       120   
...          ...       ...                                  ...       ...   
541194    581493     23204          CHARLOTTE BAG APPLES DESIGN        10   
541195    581493     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR        18   
541196    581493     22252  BIRDCAGE DECORATION TEALIGHT HOLDER        12   
541197    581493     22807         SET OF 6 T-LIGHTS TOADSTOOLS         6   
541198    581493      POST                              POSTAGE         1   

             InvoiceDate  UnitPrice  CustomerID  Country        date day  \
7279    12/03/2010 16:35       2.10     12395.0  Belgium  12/03/2010  12   
7280    12/03/2010 16:35       9.95     12395.0  Belgium  12/03/2010  12   
7281    12/03/2010 16:35       0.42     12395.0  Belgium  12/03/2010  12   
7282    12/03/2010 16:35       0.42     12395.0  Belgium  12/03/2010  12   
7283    12/03/2010 16:35       0.42     12395.0  Belgium  12/03/2010  12   
...                  ...        ...         ...      ...         ...  ..   
541194  12/09/2011 10:10       0.85     12423.0  Belgium  12/09/2011  12   
541195  12/09/2011 10:10       0.79     12423.0  Belgium  12/09/2011  12   
541196  12/09/2011 10:10       1.25     12423.0  Belgium  12/09/2011  12   
541197  12/09/2011 10:10       2.95     12423.0  Belgium  12/09/2011  12   
541198  12/09/2011 10:10      15.00     12423.0  Belgium  12/09/2011  12   

       month  year  
7279      03  2010  
7280      03  2010  
7281      03  2010  
7282      03  2010  
7283      03  2010  
...      ...   ...  
541194    09  2011  
541195    09  2011  
541196    09  2011  
541197    09  2011  
541198    09  2011  

[2069 rows x 12 columns]

### หาข้อมูลทวีปยุโรป

In [52]:
EU_data = data_test[data_test['Country']==""]
for i in range(len(UP_country)):
  x = data_test[data_test['Country']==UP_country[i]]
  EU_data=pd.concat([x,EU_data])

In [53]:
len(set(EU_data['Country']))

22

###หาข้อ 2

In [54]:
transactionsEU = EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [55]:
frequentItemsets = list(apriori(transactionsEU,min_support = 0.05))
#frequentItemsets

In [70]:
min_sup = 0.03

frequentItemsetsEU= list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03243053387449266, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03243053387449266, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6772616136919316, lift=8.201394890946151), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39272211720226846, lift=8.201394890946151)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.030049953168904152, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.030049953168904152, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7389635316698656, lift=17.213819577735123), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.21381957773

##3.ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)

##เเยกทวีปเอเชีย

In [72]:
Asia_data = data_test[data_test['Country']==""]
for i in range(len(Asia_country)):
  x = data_test[data_test['Country']==Asia_country[i]]
  Asia_data=pd.concat([x,Asia_data])

In [73]:
len(set(Asia_data['Country']))

8

In [74]:
transactionsAsia = Asia_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [75]:
frequentItemsets = list(apriori(transactionsAsia,min_support = 0.05))
#frequentItemsets

In [76]:
min_sup = 0.03

frequentItemsetsAsia= list(apriori(transactionsAsia,min_support = min_sup))
for i in frequentItemsetsAsia:
    if i[0].__len__()>1:
        print('yes')
        print(i)

KeyboardInterrupt: 

###หาข้อ 3

###5.ของวันวาเลนไทน์ (14 Feb)(ไม่มี 14 กุมพาพันธฺ์)

In [34]:
data_test[(data_test['month']=='02')]

InvoiceNo StockCode                      Description  Quantity  \
3108      536598     21421            PORCELAIN ROSE LARGE         12   
3109      536598     21422             PORCELAIN ROSE SMALL        16   
3110      536598     22178  VICTORIAN GLASS HANGING T-LIGHT        24   
3111      536598     22617       BAKING SET SPACEBOY DESIGN        24   
3112      536599     22968       ROSE COTTAGE KEEPSAKE BOX          8   
...          ...       ...                              ...       ...   
522160    580379    72225C                              NaN      -144   
522161    580380     21620                            check       -20   
522162    580381     21758                              NaN        -9   
522163    580382     16045                            check      -900   
522164    580383     20652                            check       -37   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
3108    12/02/2010 07:48       1.25     13090.0  United Kingdom  12/02/2010   
3109    12/02/2010 07:48       0.85     13090.0  United Kingdom  12/02/2010   
3110    12/02/2010 07:48       1.25     13090.0  United Kingdom  12/02/2010   
3111    12/02/2010 07:48       4.25     13090.0  United Kingdom  12/02/2010   
3112    12/02/2010 07:49       8.50     15694.0  United Kingdom  12/02/2010   
...                  ...        ...         ...             ...         ...   
522160  12/02/2011 17:57       0.00         NaN  United Kingdom  12/02/2011   
522161  12/02/2011 17:58       0.00         NaN  United Kingdom  12/02/2011   
522162  12/02/2011 17:58       0.00         NaN  United Kingdom  12/02/2011   
522163  12/02/2011 17:59       0.00         NaN  United Kingdom  12/02/2011   
522164  12/02/2011 17:59       0.00         NaN  United Kingdom  12/02/2011   

       day month  year  
3108    12    02  2010  
3109    12    02  2010  
3110    12    02  2010  
3111    12    02  2010  
3112    12    02  2010  
...     ..   ...   ...  
522160  12    02  2011  
522161  12    02  2011  
522162  12    02  2011  
522163  12    02  2011  
522164  12    02  2011  

[16729 rows x 12 columns]

#quiz 22386 เเละ 85099B คือสินค้าอะไร

In [35]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [36]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [37]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'546315',
 '545945',
 'C548485',
 '576339',
 '547788',
 '560906',
 '578943',
 '578044',
 '565784',
 '550186',
 '576181',
 '545429',
 'C566758',
 '569918',
 '563931',
 '556503',
 '544301',
 '579512',
 '559295',
 '563384',
 '536856',
 '548409',
 '541421',
 '550265',
 '536524',
 '557642',
 'C553479',
 '556953',
 '552663',
 '553385',
 '564172',
 '549746',
 '563534',
 '552323',
 '543045',
 '577691',
 '569137',
 '559052',
 '546683',
 '567368',
 '549124',
 '567631',
 'C552698',
 '580983',
 '547949',
 '548003',
 '536544',
 '572553',
 '551352',
 '547360',
 '545912',
 '551995',
 '550953',
 'C542263',
 '551160',
 '549524',
 '537955',
 '544928',
 'C551511',
 'C562953',
 '560357',
 '546837',
 '565916',
 '565396',
 '541220',
 '544668',
 '569577',
 '571216',
 '553718',
 '558862',
 '546896',
 '558777',
 '549296',
 '576899',
 '539609',
 '570187',
 '563896',
 '544338',
 'C562119',
 '556459',
 '541280',
 '575739',
 '550458',
 '573083',
 '540662',
 '556927',
 '578065',
 '548860',
 '556917',
 '553901',
 '

In [38]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  08/11/2011   
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  08/11/2011   
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  08/11/2011   
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  08/11/2011   
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  08/11/2011   
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  08/11/2011   
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  08/11/2011   
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  08/11/2011   
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  08/11/2011   
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  08/11/2011   
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  08/11/2011   
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  08/11/2011   
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  08/11/2011   
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom  08/11/2011   

       day month  year  
297488  08    11  2011  
297489  08    11  2011  
297490  08    11  2011  
297491  08    11  2011  
297492  08    11  2011  
297493  08    11  2011  
297494  08    11  2011  
297495  08    11  2011  
297496  08    11  2011  
297497  08    11  2011  
297498  08    11  2011  
297499  08    11  2011  
297500  08    11  2011  
297501  08    11  2011  
297502  08    11  2011  
297503  08    11  2011  
297504  08    11  2011  
297505  08    11  2011